kola/tests/bpf/bpf.go

```diff
@@ -0,0 +1,137 @@
+// Copyright © Microsoft Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bpf
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/coreos/mantle/kola/cluster"
+	"github.com/coreos/mantle/kola/register"
+	"github.com/coreos/mantle/util"
+	"github.com/coreos/pkg/capnslog"
+)
+
+var (
+	plog = capnslog.NewPackageLogger("github.com/coreos/mantle", "kola/tests/bpf")
+	// cmdPrefix is a temporary hack to pull `bcc` tools into Flatcar
+	cmdPrefix = "docker run -d --name %s -v /lib/modules:/lib/modules -v /sys/kernel/debug:/sys/kernel/debug -v /sys/fs/cgroup:/sys/fs/cgroup -v /sys/fs/bpf:/sys/fs/bpf --privileged --net host --pid host quay.io/iovisor/bcc %s"
+)
+
+// Log defines the standarg log format
+// from Docker
+// https://docs.docker.com/config/containers/logging/json-file/
+type Log struct {
+	Log    string `json:"log"`
+	Time   string `json:"time"`
+	Stream string `json:"stream"`
+}
+
+func init() {
+	register.Register(&register.Test{
+		Run:         execsnoopTest,
+		ClusterSize: 1,
+		Name:        `bpf.execsnoop`,
+		Distros:     []string{"cl"},
+		// required while SELinux policy is not correcly updated to support
+		// `bpf` and `perfmon` permission.
+		Flags: []register.Flag{register.NoEnableSelinux},
+	})
+}
+
+func execsnoopTest(c cluster.TestCluster) {
+	c.Run("filter name and args", func(c cluster.TestCluster) {
+		m := c.Machines()[0]
+		containerName := "execsnoop"
+
+		// filter commands with `docker ps`
+		plog.Infof("running %s container", containerName)
+		cmd := fmt.Sprintf(cmdPrefix, containerName, "/usr/share/bcc/tools/execsnoop -n docker -l ps")
+		if _, err := c.SSH(m, cmd); err != nil {
+			c.Fatalf("unable to run SSH command '%s': %v", cmd, err)
+		}
+
+		// wait for the container and the `execsnoop` command to be correctly started before
+		// generating traffic.
+		if err := util.Retry(5, 2*time.Second, func() error {
```

Comment:
Every loop of the retry runs `docker ps` so now its definitely possible to have more than 3 lines in the log (if the retry happens)?

Reply:
Mm, I get your point but it should not:

in the iteration, we run `docker ps` then we check `sudo cat $(docker inspect --format='{{.LogPath}}' %s)`:

- empty: it means `execsnoop` is not correctly started (or does not filter correctly) -> we continue to iterate
- not empty: `execsnoop` is correctly started because it generated some logs based on the `docker ps` "trigger" -> we continue the program

Even in the case we loop 5 times, it means that in the meantime not logs have been generated -> `execsnoop` was not ready. In the end, we should always get 3 lines into the logs.